SO-3746: Create script to import latest export file in specified ...

AAAlinaaa opened the pull request:
...target folder to a new task on a specified server

AAAlinaaa committed:
SO-3746: Create script to import latest export file in specified ...

...target folder to a new task on a specified server

apeteri commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,244 @@
+	check_if_empty "${SNOW_OWL_USER}" "Snow Owl username must be specified"
+	check_if_empty "${SNOW_OWL_USER_PASSWORD}" "User password must be specified"
+	check_if_empty "${SOURCE_FOLDER}" "Source folder must be specified"
+	SOURCE_FILE=$(ls -t snow_owl_DELTA_export* | head -1)
```
For clarity, please replace the `-1` option with `-n1`; omitting the 'n' is a deprecated option according to the manual page (last paragraph of https://www.gnu.org/software/coreutils/manual/html_node/head-invocation.html):

> For compatibility head also supports an obsolete option syntax -[NUM][bkm][cqv], which is recognized only if it is specified first.

apeteri commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,244 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
```
`set -euo pipefail` is a common default incantation at the start of bash scripts: http://redsymbol.net/articles/unofficial-bash-strict-mode/ 

Short summary: 
* `set -e` exits the script if a command returns a non-zero exit code, doesn't move on to the next line in the script;
* `set -u` exits the script if a variable is referenced which is undefined; 
* `set -o pipefail` means the exit code from piped together processes only returns success if all processes in the pipe succeeded (default is only the last process status counts)

If it doesn't break any functionality, I'd recommend adding it here.

apeteri commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,244 @@
+
+bugzilla_login() {
+	BUGZILLA_LOGIN_ENDPOINT="${BUGZILLA_URL}/rest/login?login=${BUGZILLA_USERNAME}&password=${BUGZILLA_PASSWORD}"
+	BUGZILLA_TOKEN=$(curl --request GET \
```
Adding `--fail` to all invocations of curl would make it return a non-zero exit status if a 4xx or 5xx response is encountered. In combination with the `e` option of  `set -euo pipefail` mentioned above, the script would stop at that point and not try to make the rest of the HTTP calls.

https://superuser.com/a/1249678

apeteri commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,244 @@
+
+bugzilla_lougout() {
+	BUGZILLA_LOGOUT_ENDPOINT="${BUGZILLA_URL}/rest/logout?token=${BUGZILLA_TOKEN}"
+	BUGZILLA_LOGOUT_ENDPOINT="$(echo -e "${BUGZILLA_LOGOUT_ENDPOINT}" | tr -d '[:space:]')"
```
Is this filtering step still needed? It is only present in `bugzilla_logout` and `create_bug` but not in `bugzilla_login`.

nagyo commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,244 @@
+        Snow Owl base URL, defaults to 'http://localhost:8080'
+    -f
+        Source folder where the exported content can be found
+    -s
```
This option does not exist among the available parameters. Please remove this line!

nagyo commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,244 @@
+	-U
+		Defines the bugzilla user
+	-P
+		Defines the bugzilla pasword for the above user
```
Typo in password

nagyo commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,244 @@
+	check_if_empty "${SNOW_OWL_USER_PASSWORD}" "User password must be specified"
+	check_if_empty "${SOURCE_FOLDER}" "Source folder must be specified"
+	SOURCE_FILE=$(ls -t snow_owl_DELTA_export* | head -1)
+	check_if_empty "${SOURCE_FILE}" "Source folder must contain a valid delta export file"
```
The check_if_empty function validates if SOURCE_FILE is an empty string or not. You can also validate if it points to a valid file by using the -f parameter:
```
FILE=/path/to/file.txt
if [ -f "$FILE" ]; then
    echo "$FILE exist"
fi
```

nagyo reviewed (changes requested):
Added a few low severity comments. Beside that I'm seeing that the following requirement is not fulfilled yet?

> The script should be able to specify more than one server as the target

AAAlinaaa committed:
SO-3746: Enable importing on more than one server, ensure script ...

... terminates early if a routine returns with a non-zero exit code,
remove unused branch parameter from help, exclude headers from login
response for cleaner bugzilla token parsing

nagyo commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -214,7 +222,7 @@ while getopts ":hu:p:f:b:a:z:U:P:" option; do
 		;;
 	b)
-		SNOW_OWL_BASE_URL=${OPTARG}
+		SNOW_OWL_BASE_URLS=${OPTARG}
```
`getops` allows multi valued parameters, the only restriction is that the user needs to write the parameter letter before each possible value. E.g. **-u** http://snowowl-1.url **-u** http://snowowl-2.url
I would recommend using this solution compared to the space separated list of URLs. It will make it much more clean and terminal friendly! Here is a short example:
```
URLS=()

for URL in "${URLS[@]}"; do
	echo "${URL}"
done

while getopts "u:" opt; do
	case "$opt" in
	u)
		URLS+=("$OPTARG")
		;;
	esac
done
```

AAAlinaaa committed:
SO-3746: Read in target servers as multi-valued parameters instead ...

... of a single space separated string

nagyo commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,260 @@
+SNOW_OWL_BASE_URLS=()
+
+#Snow Owl server base url from the target servers currently undergoing import
+SNOW_OWL_BASE_URL="http://localhost:8080"
```
Could you please rename this to DEFAULT_SNOW_OWL_BASE_URL so this will be distinguished with the variable introduced in the final for loop.

nagyo commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,260 @@
+    -p
+        Define the password for the above user
+    -b
+        Space separarated list of Snow Owl base URLs, defaults to 'http://localhost:8080'
```
This text needs some updating as well!

nagyo commented on releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_delta_import.sh:
```diff
@@ -0,0 +1,260 @@
+		- Bugzilla URL
+		- Bugzilla user
+		- Bugzilla password for the above user
+
```
At least one simple example would be nice here, see https://github.com/b2ihealthcare/snow-owl/blob/6.x/releng/com.b2international.snowowl.server.update/assembly/common/scripts/snowowl_hot_backup_mysql.sh#L106

AAAlinaaa committed:
SO-3746: Update variable names, and their descriptions
